File: cpp/3737_count_majority_subarrays/count_majority_subarrays.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int countMajoritySubarrays(vector<int>& nums, int target) {
    int n = nums.size();
    int ans = 0;
    for (int i = 0; i < n; ++i) {
      int cnt = 0;
      for (int j = i; j < n; ++j) {
        cnt += (nums[j] == target ? 1 : -1);
        if (cnt > 0) {
          ++ans;
        }
      }
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int, int>> tests{
      {{1, 2, 2, 3}, 2, 5},
      {{1, 1, 1, 1}, 1, 10},
      {{1, 2, 3}, 4, 0},
  };

  for (auto& [nums, target, expected] : tests) {
    assert(Solution().countMajoritySubarrays(nums, target) == expected);
  }
}

File: rust/3737_count_majority_subarrays/count_majority_subarrays.rs
struct Solution;

impl Solution {
    pub fn count_majority_subarrays(nums: Vec<i32>, target: i32) -> i32 {
        let n = nums.len();
        let mut ans = 0;
        for i in 0..n {
            let mut cnt = 0;
            for j in i..n {
                cnt += if nums[j] == target { 1 } else { -1 };
                if cnt > 0 {
                    ans += 1;
                }
            }
        }
        ans
    }
}

fn main() {
    let tests = vec![
        (vec![1, 2, 2, 3], 2, 5),
        (vec![1, 1, 1, 1], 1, 10),
        (vec![1, 2, 3], 4, 0),
    ];

    for (nums, target, expected) in tests {
        assert_eq!(Solution::count_majority_subarrays(nums, target), expected);
    }
}
